src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -790,11 +790,15 @@ HRESULT CorProfiler::ProcessReplacementCalls(
         continue;
       }
 
+      bool caller_assembly_is_domain_neutral =
+          runtime_information_.is_desktop() && corlib_module_loaded &&
+          module_metadata->app_domain_id == corlib_app_domain_id;
+
       // At this point we know we've hit a match. Error out if
-      //   1) The target assembly is Datadog.Trace.ClrProfiler.Managed
-      //   2) The managed profiler has not been loaded yet
+      //   1) The managed profiler has not been loaded yet
+      //   2) The caller is domain-neutral AND we want to instrument domain-neutral assemblies AND the Profiler has already been loaded
       if (!ProfilerAssemblyIsLoadedIntoAppDomain(module_metadata->app_domain_id) &&
-          method_replacement.wrapper_method.assembly.name == "Datadog.Trace.ClrProfiler.Managed"_W) {
+          !(caller_assembly_is_domain_neutral && instrument_domain_neutral_assemblies && ProfilerAssemblyIsLoadedIntoAnyAppDomain())) {
```

Comment:
This was kind of tough for my brain to read, but it checks out. Not sure how to simplify the reading outright.

Reply:
Yeah it was tough to assemble and understand, and I couldn't figure out a better way to simplify it